File: library/cpp/sandesh_session.cc
/*
 * Copyright (c) 2013 Juniper Networks, Inc. All rights reserved.
 */

//
// sandesh_session.cc
//
// Sandesh session
//

#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN
#endif

#include <boost/bind.hpp>
#include <boost/assign.hpp>
#include <boost/algorithm/string.hpp>

#include <base/parse_object.h>

#include <sandesh/common/vns_types.h>
#include <sandesh/common/vns_constants.h>
#include <sandesh/transport/TBufferTransports.h>
#include <sandesh/protocol/TXMLProtocol.h>
#include "sandesh/sandesh_types.h"
#include "sandesh/sandesh.h"

#include "sandesh_connection.h"
#include "sandesh_session.h"


using namespace std;
using namespace contrail::sandesh::protocol;
using namespace contrail::sandesh::transport;

using boost::asio::mutable_buffer;
using boost::asio::buffer_cast;

const std::string SandeshWriter::sandesh_open_ = sXML_SANDESH_OPEN;
const std::string SandeshWriter::sandesh_open_attr_length_ =
        sXML_SANDESH_OPEN_ATTR_LENGTH;
const std::string SandeshWriter::sandesh_close_ = sXML_SANDESH_CLOSE;

//
// SandeshWriter
//
SandeshWriter::SandeshWriter(SandeshSession *session)
    : session_(session),
    ready_to_send_(true),
    send_buf_(new uint8_t[kDefaultSendSize]),
    send_buf_offset_(0) {
}

SandeshWriter::~SandeshWriter() {
    delete [] send_buf_;
}

void SandeshWriter::WriteReady(const boost::system::error_code &ec) {
    if (ec) {
        SANDESH_LOG(ERROR, "SandeshSession Write error value: " << ec.value()
            << " category: " << ec.category().name()
            << " message: " << ec.message());
        session_->increment_write_ready_cb_error();
        return;
    }

    {
        tbb::mutex::scoped_lock lock(send_mutex_);
        ready_to_send_ = true;
    }

    // We may want to start the Runner for the send_queue
    session_->send_queue()->MayBeStartRunner();
}

void SandeshWriter::SendMsg(Sandesh *sandesh, bool more) {
    SandeshHeader header;
    std::stringstream ss;
    uint8_t *buffer;
    int32_t xfer = 0, ret;
    uint32_t offset;
    boost::shared_ptr<TMemoryBuffer> btrans(
                    new TMemoryBuffer(kEncodeBufferSize));
    boost::shared_ptr<TXMLProtocol> prot(
                    new TXMLProtocol(btrans));
    // Populate the header
    header.set_Namespace(sandesh->scope());
    header.set_Timestamp(sandesh->timestamp());
    header.set_Module(sandesh->module());
    header.set_Source(sandesh->source());
    header.set_Context(sandesh->context());
    header.set_SequenceNum(sandesh->seqnum());
    header.set_VersionSig(sandesh->versionsig());
    header.set_Type(sandesh->type());
    header.set_Hints(sandesh->hints());
    header.set_Level(sandesh->level());
    header.set_Category(sandesh->category());
    header.set_NodeType(sandesh->node_type());
    header.set_InstanceId(sandesh->instance_id());

    // Write the sandesh open envelope.
    buffer = btrans->getWritePtr(sandesh_open_.length());
    memcpy(buffer, sandesh_open_.c_str(), sandesh_open_.length());
    btrans->wroteBytes(sandesh_open_.length());
    // Write the sandesh header
    if ((ret = header.write(prot)) < 0) {
        SANDESH_LOG(ERROR, __func__ << ": Sandesh header write FAILED: " <<
            sandesh->Name() << " : " << sandesh->source() << ":" <<
            sandesh->module() << ":" << sandesh->instance_id() <<
            " Sequence Number:" << sandesh->seqnum());
        session_->increment_send_msg_fail();
        Sandesh::UpdateTxMsgFailStats(sandesh->Name(), 0,
            SandeshTxDropReason::HeaderWriteFailed);
        sandesh->Release();
        return;
    }
    xfer += ret;
    // Write the sandesh
    if ((ret = sandesh->Write(prot)) < 0) {
        SANDESH_LOG(ERROR, __func__ << ": Sandesh write FAILED: "<<
            sandesh->Name() << " : " << sandesh->source() << ":" <<
            sandesh->module() << ":" << sandesh->instance_id() <<
            " Sequence Number:" << sandesh->seqnum());
        session_->increment_send_msg_fail();
        Sandesh::UpdateTxMsgFailStats(sandesh->Name(), 0,
            SandeshTxDropReason::WriteFailed);
        sandesh->Release();
        return;
    }
    xfer += ret;
    // Write the sandesh close envelope
    buffer = btrans->getWritePtr(sandesh_close_.length());
    memcpy(buffer, sandesh_close_.c_str(), sandesh_close_.length());
    btrans->wroteBytes(sandesh_close_.length());
    // Get the buffer
    btrans->getBuffer(&buffer, &offset);
    // Sanity
    assert(sandesh_open_.length() + xfer + sandesh_close_.length() ==
            offset);
    // Update the sandesh open envelope length;
    char prev = ss.fill('0');
    // Adjust for '">'
    ss.width(sandesh_open_.length() - sandesh_open_attr_length_.length() - 2);
    ss << offset;
    ss.fill(prev);
    memcpy(buffer + sandesh_open_attr_length_.length(), ss.str().c_str(),
            ss.str().length());

    // Update sandesh stats
    Sandesh::UpdateTxMsgStats(sandesh->Name(), offset);
    session_->increment_send_msg();

    if (send_buf()) {
        if (more) {
            // There are more messages in the send_queue_. 
            // Try to package as many sandesh messages as possible 
            // (== kEncodeBufferSize) before transporting to the
            // receiver.
            SendMsgMore(btrans);
        } else {
            // send_queue_ is empty. Flush sandesh->send_buf_ and this message.
            SendMsgAll(btrans);
        }
    } else {
        // Send the message
        SendInternal(btrans);
    }
    sandesh->Release();
}

// Package as many sandesh messages as possible [not more than 
// kEncodeBufferSize] before transporting it to the receiver. 
// 
// send_buf_ => unsent data (partial/complete message).
// buf => new message
// buf_len => buf's len
void SandeshWriter::SendMsgMore(boost::shared_ptr<TMemoryBuffer>
                                            send_buffer) {
    uint8_t *buf;
    uint32_t buf_len;

    send_buffer->getBuffer(&buf, &buf_len);

    if (send_buf_offset()) {
        // We have some unsent data.
        size_t bulk_msg_len = send_buf_offset() + buf_len;
        if (bulk_msg_len < kDefaultSendSize) {
            // We still have space for more data. Don't send the message yet.
            // Add the message to the existing data.
            append_send_buf(buf, buf_len);
        } else {
            uint8_t *buffer;
            // (send_buf_offset() + buf_len) >= kDefaultSendSize
            boost::shared_ptr<TMemoryBuffer> bulk_msg(
                    new TMemoryBuffer(send_buf_offset()));
            buffer = bulk_msg->getWritePtr(send_buf_offset());
            // Copy unsent data
            memcpy(buffer, send_buf(), send_buf_offset());
            bulk_msg->wroteBytes(send_buf_offset());
            // Send it
            SendInternal(bulk_msg);
            // Cleanup send_buf
            reset_send_buf();
            // Next send the new message
            SendInternal(send_buffer);
        }
    } else {
        // We don't have any unsent data.
        if (buf_len >= kDefaultSendSize) {
            // We don't have room to accommodate anything more.
            // Send the message now.
            SendInternal(send_buffer);
        } else {
            // We have room to accommodate more data.
            // Save this message.
            // Note: The memcpy here can be avoided by passing send_buf_ 
            // to TMemoryBuffer so that the message is encoded in 
            // send_buf_ itself. 
            set_send_buf(buf, buf_len);
        }
    }
}

// sandesh->send_queue_ is empty. 
// Flush unsent data (if any) and this message. 
void SandeshWriter::SendMsgAll(boost::shared_ptr<TMemoryBuffer> send_buffer) {
    uint8_t *buf;
    uint32_t buf_len;

    send_buffer->getBuffer(&buf, &buf_len);

    if (send_buf_offset()) {
        // We have some unsent data.
        size_t bulk_msg_len = send_buf_offset() + buf_len;
        if (bulk_msg_len <= kDefaultSendSize) {
            uint8_t *buffer;
            // We have enough room to send all the pending data in one message.
            boost::shared_ptr<TMemoryBuffer> bulk_msg(new 
                    TMemoryBuffer(bulk_msg_len));
            buffer = bulk_msg->getWritePtr(bulk_msg_len);
            memcpy(buffer, send_buf(), send_buf_offset());
            memcpy(buffer + send_buf_offset(), buf, buf_len);
            bulk_msg->wroteBytes(bulk_msg_len);
            // send the message
            SendInternal(bulk_msg);
            // reset send_buf_
            reset_send_buf();
        } else {
            uint8_t *buffer;
            // We don't have enough space to accommodate all the
            // pending data in one message. 
            boost::shared_ptr<TMemoryBuffer> old_buf(new
                    TMemoryBuffer(send_buf_offset()));
            buffer = old_buf->getWritePtr(send_buf_offset());
            memcpy(buffer, send_buf(),
                   send_buf_offset());
            old_buf->wroteBytes(send_buf_offset());
            // Take care of the unsent data in send_buf_ first.
            // Note that we could have accomodated part of buf [last message] 
            // here. But, not doing so to avoid additional memcpy() :)
            SendInternal(old_buf);
            // Cleanup send_buf_
            reset_send_buf();
            // Well, send the last message now.
            SendInternal(send_buffer);
        }
    } else {
        // No unsent data. Send the message now.
        SendInternal(send_buffer);
    }
}

void SandeshWriter::SendInternal(boost::shared_ptr<TMemoryBuffer> buf) {
    uint8_t  *buffer;
    uint32_t len;
    buf->getBuffer(&buffer, &len);
    tbb::mutex::scoped_lock lock(send_mutex_);
    ready_to_send_ = session_->Send((const uint8_t *)buffer, len, NULL);
}

//
// SandeshSession
//
SandeshSession::SandeshSession(SslServer *client, SslSocket *socket,
        int task_instance, int writer_task_id, int reader_task_id) :
    SslSession(client, socket),
    instance_(task_instance),
    writer_(new SandeshWriter(this)), 
    reader_(new SandeshReader(this)),
    send_queue_(new Sandesh::SandeshQueue(writer_task_id,
            task_instance,
            boost::bind(&SandeshSession::SendMsg, this, _1),
            kQueueSize)),
    keepalive_idle_time_(kSessionKeepaliveIdleTime),
    keepalive_interval_(kSessionKeepaliveInterval),
    keepalive_probes_(kSessionKeepaliveProbes),
    tcp_user_timeout_(kSessionTcpUserTimeout),
    reader_task_id_(reader_task_id) {
    if (Sandesh::role() == Sandesh::SandeshRole::Collector) {
        send_buffer_queue_.reset(new Sandesh::SandeshBufferQueue(writer_task_id,
                task_instance,
                boost::bind(&SandeshSession::SendBuffer, this, _1)));
        send_buffer_queue_->SetStartRunnerFunc(boost::bind(&SandeshSession::SessionSendReady, this));
    }
    send_queue_->SetStartRunnerFunc(boost::bind(&SandeshSession::SessionSendReady, this));
}

SandeshSession::~SandeshSession() {
}

bool SandeshSession::SessionSendReady() {
    return (IsEstablished() && writer_->SendReady() &&
            Sandesh::IsSendQueueEnabled());
}

void SandeshSession::SetSendQueueWaterMark(
    Sandesh::QueueWaterMarkInfo &swmi) {
    WaterMarkInfo wm(boost::get<0>(swmi),
        boost::bind(&Sandesh::SetSendingLevel, _1, boost::get<1>(swmi)));
    if (boost::get<2>(swmi)) {
        send_queue_->SetHighWaterMark(wm);
    } else {
        send_queue_->SetLowWaterMark(wm);
    }
}

void SandeshSession::ResetSendQueueWaterMark() {
    send_queue_->ResetHighWaterMark();
    send_queue_->ResetLowWaterMark();
}

void SandeshSession::Shutdown() {
    if (Sandesh::role() == Sandesh::SandeshRole::Collector) {
        send_buffer_queue_->Shutdown();
    }
    send_queue_->Shutdown();
}

std::string SandeshSession::ToString() const {
    std::stringstream out;
    out << TcpSession::ToString() << "(" << instance_ << ")";
    return out.str();
}

boost::system::error_code SandeshSession::SetSocketOptions() {
    boost::system::error_code ec = TcpSession::SetSocketOptions();
    if (ec) {
        return ec;
    }
    return SetSocketKeepaliveOptions(keepalive_idle_time_, keepalive_interval_,
            keepalive_probes_, tcp_user_timeout_);
}

void SandeshSession::OnRead(Buffer buffer) {
    reader_->OnRead(buffer);
}

bool SandeshSession::SendMsg(SandeshElement element) {
    Sandesh *sandesh = element.snh_;
    tbb::mutex::scoped_lock lock(send_mutex_);
    if (!IsEstablished()) {
        if (Sandesh::IsLoggingDroppedAllowed(sandesh->type())) {
            SANDESH_LOG(ERROR, __func__ << " Not Connected : Dropping Message: " <<
                sandesh->ToString());
        }
        increment_send_msg_fail();
        Sandesh::UpdateTxMsgFailStats(sandesh->Name(), 0,
            SandeshTxDropReason::SessionNotConnected);
        sandesh->Release();
        return true;
    }
    if (sandesh->IsLoggingAllowed()) {
        sandesh->Log();
    }
    bool more = !send_queue_->IsQueueEmpty();
    writer_->SendMsg(sandesh, more);
    return true;
}

bool SandeshSession::SendBuffer(boost::shared_ptr<TMemoryBuffer> sbuffer) {
    tbb::mutex::scoped_lock lock(send_mutex_);
    if (!IsEstablished()) {
        increment_send_buffer_fail();
        return true;
    }
    // No buffer packing supported currently
    writer_->SendBuffer(sbuffer);
    return true;
}

bool SandeshSession::EnqueueBuffer(u_int8_t *buf, u_int32_t buf_len) {
    boost::shared_ptr<TMemoryBuffer> sbuffer(new TMemoryBuffer(buf_len));
    u_int8_t *write_buf = sbuffer->getWritePtr(buf_len);
    memcpy(write_buf, buf, buf_len);
    sbuffer->wroteBytes(buf_len);
    return send_buffer_queue()->Enqueue(sbuffer);
}

Sandesh * SandeshSession::DecodeCtrlSandesh(const string& msg,
        const SandeshHeader& header,
        const string& sandesh_name, const uint32_t& header_offset) {
    namespace sandesh_prot = contrail::sandesh::protocol;
    namespace sandesh_trans = contrail::sandesh::transport;

    assert(header.get_Hints() & g_sandesh_constants.SANDESH_CONTROL_HINT);

    // Create and process the sandesh
    Sandesh *sandesh = SandeshBaseFactory::CreateInstance(sandesh_name);
    if (sandesh == NULL) {
        SANDESH_LOG(ERROR, __func__ << ": Unknown sandesh ctrl message: " << sandesh_name);
        return NULL;
    }
    boost::shared_ptr<sandesh_trans::TMemoryBuffer> btrans =
            boost::shared_ptr<sandesh_trans::TMemoryBuffer>(
                    new sandesh_trans::TMemoryBuffer((uint8_t *)msg.c_str() + header_offset,
                            msg.size() - header_offset));
    boost::shared_ptr<sandesh_prot::TXMLProtocol> prot =
            boost::shared_ptr<sandesh_prot::TXMLProtocol>(new sandesh_prot::TXMLProtocol(btrans));
    int32_t xfer = sandesh->Read(prot);
    if (xfer < 0) {
        SANDESH_LOG(ERROR, __func__ << ": Decoding " << sandesh_name << " for ctrl FAILED");
        sandesh->Release();
        return NULL;
    } else {
        return sandesh;
    }
}

void SandeshSession::EnqueueClose() {
    if (IsClosed()) {
        return;
    }
    tbb::mutex::scoped_lock lock(conn_mutex_);
    if (connection_) {
        connection_->state_machine()->OnSessionEvent(this,
            TcpSession::CLOSE);
    }
}

//
// SandeshReader
//
SandeshReader::SandeshReader(SandeshSession *session) :
        buf_(""),
        offset_(0),
        msg_length_(-1),
        session_(session) {
    buf_.reserve(kDefaultRecvSize);
}

SandeshReader::~SandeshReader() {
}

int SandeshReader::ExtractMsgHeader(const std::string& msg,
        SandeshHeader& header, std::string& msg_type, uint32_t& header_offset) {
    int32_t xfer = 0, ret;
    boost::shared_ptr<TMemoryBuffer> btrans =
            boost::shared_ptr<TMemoryBuffer>(
                    new TMemoryBuffer((uint8_t *)msg.c_str(), msg.size()));
    boost::shared_ptr<TXMLProtocol> prot =
            boost::shared_ptr<TXMLProtocol>(new TXMLProtocol(btrans));
    // Read the sandesh header and note the offset
    if ((ret = header.read(prot)) <= 0) {
        SANDESH_LOG(ERROR, __func__ << ": Sandesh header read FAILED: " << msg);
        return EINVAL;
    }
    xfer += ret;
    header_offset = xfer;
    // Extract the message name
    if ((ret = prot->readSandeshBegin(msg_type)) <= 0) {
        SANDESH_LOG(ERROR, __func__ << ": Sandesh begin read FAILED: " << msg);
        return EINVAL;
    }
    xfer += ret;
    return 0;
}

void SandeshReader::SetBuf(const std::string &str) {
    if (buf_.empty()) {
        ReplaceBuf(str);
    } else {
        buf_ += str;
    }
    // TODO handle buf_ > kMaxMessageSize
}

void SandeshReader::ReplaceBuf(const std::string &str) {
    buf_ = str;
    buf_.reserve(SandeshReader::kDefaultRecvSize);
    offset_ = 0;
}

bool SandeshReader::LeftOver() const {
    if (buf_.empty()) {
        return false;
    }
    return (buf_.size() != offset_);
}

// Returns false if not able to extract the message length, true otherwise
bool SandeshReader::ExtractMsgLength(size_t &msg_length, int *result) {
    // Have we read enough to extract the message length?
    if (buf_.size() - offset_ < SandeshWriter::sandesh_open_.size()) {
        return false;
    }
    // Some sanity check
    if (!boost::algorithm::starts_with(buf_.c_str() + offset_,
            SandeshWriter::sandesh_open_attr_length_)) {
        *result = -1;
        return false;
    }

    std::string::const_iterator end = buf_.begin() + offset_ +
            SandeshWriter::sandesh_open_.size() - 1;
    if (*end != '>') {
        *result = -2;
        return false;
    }

    std::string::const_iterator st = buf_.begin() + offset_ +
            SandeshWriter::sandesh_open_attr_length_.size();
    // Adjust for double quote
    --end;
    string length = string(st, end);

    stringToInteger(length.c_str(), msg_length);
    if (msg_length == 0) {
	*result = -3;
	return false;
    }
    return true;
}

// Returns false if not able to extract the full message, true otherwise
bool SandeshReader::ExtractMsg(Buffer buffer, int *result, bool NewBuf) {
    if (NewBuf) {
        const uint8_t *cp = TcpSession::BufferData(buffer);
        // TODO Avoid this copy
        std::string str(cp, cp + TcpSession::BufferSize(buffer));
        SetBuf(str);
    }
    // Extract the message length
    if (!MsgLengthKnown()) {
        size_t msg_length = 0;
        bool done = ExtractMsgLength(msg_length, result);
        if (done == false) {
            return false;
        }
        set_msg_length(msg_length);
    }
    // Check if the entire message is read or not
    if (buf_.size() < msg_length()) {
        return false;
    }
    return true;
}

void SandeshReader::OnRead(Buffer buffer) {
    tbb::mutex::scoped_lock lock(cb_mutex_);
    // Check if session is being deleted, then drop the packet
    if (cb_.empty()) {
        SANDESH_LOG(ERROR, __func__ <<
            " Session being deleted: Dropping Message");
        session_->increment_recv_fail();
        session_->ReleaseBuffer(buffer);
        return;
    }
    int result = 0;
    bool done = ExtractMsg(buffer, &result, true);
    do {
        if (result < 0) {
            // Generate error and close connection
            SANDESH_LOG(ERROR, __func__ << " Message extract failed: " << result);
            const uint8_t *cp = TcpSession::BufferData(buffer);
            size_t cp_size = TcpSession::BufferSize(buffer);
            SANDESH_LOG(ERROR, __func__ << " OnRead Buffer Size: " << cp_size);
            SANDESH_LOG(ERROR, __func__ << " OnRead Buffer: ");
            std::string debug((const char*)cp, cp_size);
            SANDESH_LOG(ERROR, debug);
            SANDESH_LOG(ERROR, __func__ << " Reader Size: " << buf_.size());
            SANDESH_LOG(ERROR, __func__ << " Reader Offset: " << offset_);
            SANDESH_LOG(ERROR, __func__ << " Reader Buffer: " << buf_);
            buf_.clear();
            offset_ = 0;
            // Enqueue a close on the state machine
            session_->increment_recv_fail();
            session_->EnqueueClose();
            break;
        }
        if (done == true) {
            // We got good match. Process the message after extracting out
            // the sandesh open and close envelope
            std::string::const_iterator st = buf_.begin() + offset_ +
                    SandeshWriter::sandesh_open_.size();
            std::string::const_iterator end = buf_.begin() + offset_ +
                    msg_length() - SandeshWriter::sandesh_close_.size();
            std::string xml(st, end);
            offset_ += msg_length();
            reset_msg_length();
            if (!cb_(xml, session_)) {
                // Enqueue a close on the state machine
                session_->increment_recv_fail();
                session_->EnqueueClose();
                break;
            }
        } else {
            // Read more data.
            break;
        }

        if (LeftOver()) {
            ReplaceBuf(string(buf_, offset_, buf_.size() - offset_));
            done = ExtractMsg(buffer, &result, false);
        } else {
            // No more data in the Buffer
            buf_.clear();
            offset_ = 0;
            break;
        }
    } while (true);

    session_->ReleaseBuffer(buffer);
    return;
}

void SandeshReader::SetReceiveMsgCb(SandeshReceiveMsgCb cb) {
    tbb::mutex::scoped_lock lock(cb_mutex_);
    cb_ = cb;
}

File: library/python/pysandesh/sandesh_session.py
#
# Copyright (c) 2013 Juniper Networks, Inc. All rights reserved.
#

#
# Sandesh Session
#

import socket
from transport import TTransport
from protocol import TXMLProtocol
from work_queue import WorkQueue
from tcp_session import TcpSession
from sandesh_logger import SandeshLogger

_XML_SANDESH_OPEN = '<sandesh length="0000000000">'
_XML_SANDESH_OPEN_ATTR_LEN = '<sandesh length="'
_XML_SANDESH_OPEN_END = '">'
_XML_SANDESH_CLOSE = '</sandesh>'


class SandeshReader(object):

    _READ_OK = 0
    _READ_ERR = -1

    def __init__(self, session, sandesh_msg_handler):
        self._session = session
        self._sandesh_instance = session.sandesh_instance()
        self._sandesh_msg_handler = sandesh_msg_handler
        self._read_buf = ''
        self._sandesh_len = 0
        self._logger = session._logger
    # end __init__

    # Public functions

    def read_msg(self, rcv_buf):
        self._read_buf += rcv_buf
        while True:
            (ret, sandesh) = self._extract_sandesh()
            if ret < 0:
                self._logger.error('Failed to extract sandesh')
                return self._READ_ERR
            if not sandesh:
                # read more data
                self._logger.debug(
                    'Not enough data to extract sandesh. Read more data.')
                break
            # Call sandesh message handler
            self._sandesh_msg_handler(self._session, sandesh)
            self._read_buf = self._read_buf[self._sandesh_len:]
            self._sandesh_len = 0
            if not len(self._read_buf):
                break

        return self._READ_OK
    # end read_msg

    @staticmethod
    def extract_sandesh_header(sandesh_xml):
        transport = TTransport.TMemoryBuffer(sandesh_xml)
        protocol_factory = TXMLProtocol.TXMLProtocolFactory()
        protocol = protocol_factory.getProtocol(transport)

        from gen_py.sandesh.ttypes import SandeshHeader
        hdr = SandeshHeader()
        hdr_len = hdr.read(protocol)
        if hdr_len == -1:
            return (None, 0, None)
        # Extract the sandesh name
        (length, sandesh_name) = protocol.readSandeshBegin()
        if length == -1:
            return (hdr, hdr_len, None)
        return (hdr, hdr_len, sandesh_name)
    # end extract_sandesh_header

    # Private functions

    def _extract_sandesh(self):
        if not self._sandesh_len:
            (ret, length) = self._extract_sandesh_len()
            if ret < 0:
                return (self._READ_ERR, None)
            elif not length:
                return (self._READ_OK, None)
            self._sandesh_len = length
        if len(self._read_buf) < self._sandesh_len:
            return (self._READ_OK, None)
        # Sanity check
        sandesh_close_tag = self._read_buf[
            self._sandesh_len - len(_XML_SANDESH_CLOSE):self._sandesh_len]
        if sandesh_close_tag != _XML_SANDESH_CLOSE:
            return (self._READ_ERR, None)

        # Extract sandesh
        sandesh_begin = len(_XML_SANDESH_OPEN)
        sandesh_end = self._sandesh_len - len(_XML_SANDESH_CLOSE)
        sandesh = self._read_buf[sandesh_begin:sandesh_end]
        return (self._READ_OK, sandesh)
    # end _extract_sandesh

    def _extract_sandesh_len(self):
        # Do we have enough data to extract the sandesh length?
        if len(self._read_buf) < len(_XML_SANDESH_OPEN):
            self._logger.debug('Not enough data to extract sandesh length')
            return (self._READ_OK, 0)
        # Sanity checks
        if self._read_buf[:len(_XML_SANDESH_OPEN_ATTR_LEN)] != \
                _XML_SANDESH_OPEN_ATTR_LEN:
            return (self._READ_ERR, 0)
        if self._read_buf[len(_XML_SANDESH_OPEN) - len(_XML_SANDESH_OPEN_END):
                          len(_XML_SANDESH_OPEN)] != _XML_SANDESH_OPEN_END:
            return (self._READ_ERR, 0)
        len_str = self._read_buf[len(_XML_SANDESH_OPEN_ATTR_LEN):
                                 len(_XML_SANDESH_OPEN) -
                                 len(_XML_SANDESH_OPEN_END)]
        try:
            length = int(len_str)
        except ValueError:
            self._logger.error(
                'Invalid sandesh length [%s] in the received message' %
                (len_str))
            return (self._READ_ERR, 0)

        self._logger.debug('Extracted sandesh length: %s' % (len_str))
        return (self._READ_OK, length)
    # end _extract_sandesh_len

# end class SandeshReader


class SandeshWriter(object):

    _MAX_SEND_BUF_SIZE = 4096

    def __init__(self, session):
        self._session = session
        self._sandesh_instance = session.sandesh_instance()
        self._send_buf_cache = ''
        self._logger = session._logger
    # end __init__

    # Public functions

    @staticmethod
    def encode_sandesh(sandesh, sandesh_instance=None):
        transport = TTransport.TMemoryBuffer()
        protocol_factory = TXMLProtocol.TXMLProtocolFactory()
        protocol = protocol_factory.getProtocol(transport)

        from gen_py.sandesh.ttypes import SandeshHeader

        sandesh_hdr = SandeshHeader(sandesh.scope(),
                                    sandesh.timestamp(),
                                    sandesh.module(),
                                    sandesh.source_id(),
                                    sandesh.context(),
                                    sandesh.seqnum(),
                                    sandesh.versionsig(),
                                    sandesh.type(),
                                    sandesh.hints(),
                                    sandesh.level(),
                                    sandesh.category(),
                                    sandesh.node_type(),
                                    sandesh.instance_id())
        # write the sandesh header
        if sandesh_hdr.write(protocol) < 0:
            if sandesh_instance is not None:
                sandesh_instance.msg_stats().update_tx_drop_stats(
                    sandesh.__class__.__name__, 0)
            return None
        # write the sandesh
        if sandesh.write(protocol) < 0:
            if sandesh_instance is not None:
                sandesh_instance.msg_stats().update_tx_drop_stats(
                    sandesh.__class__.__name__, 0)
            return None
        # get the message
        msg = transport.getvalue()
        # calculate the message length
        msg_len = len(_XML_SANDESH_OPEN) + len(msg) + len(_XML_SANDESH_CLOSE)
        len_width = len(_XML_SANDESH_OPEN) - \
            (len(_XML_SANDESH_OPEN_ATTR_LEN) + len(_XML_SANDESH_OPEN_END))
        # pad the length with leading 0s
        len_str = (str(msg_len)).zfill(len_width)
        encoded_buf = _XML_SANDESH_OPEN_ATTR_LEN + len_str + \
            _XML_SANDESH_OPEN_END + msg + _XML_SANDESH_CLOSE
        return encoded_buf
    # end encode_sandesh

    def send_msg(self, sandesh, more):
        send_buf = self.encode_sandesh(sandesh)
        if send_buf is None:
            self._logger.error('Failed to send sandesh')
            return -1
        # update sandesh tx stats
        self._sandesh_instance.msg_stats().update_tx_stats(
            sandesh.__class__.__name__, len(send_buf))
        if more:
            self._send_msg_more(send_buf)
        else:
            self._send_msg_all(send_buf)
        return 0
    # end send_msg

    # Private functions

    def _send_msg_more(self, send_buf):
        self._send_buf_cache += send_buf
        if len(self._send_buf_cache) >= self._MAX_SEND_BUF_SIZE:
            # send the message
            self._send(self._send_buf_cache)
            # reset the cache
            self._send_buf_cache = ''
    # end _send_msg_more

    def _send_msg_all(self, send_buf):
        # send the message
        self._send(self._send_buf_cache + send_buf)
        # reset the cache
        self._send_buf_cache = ''
    # end _send_msg_all

    def _send(self, send_buf):
        if self._session.write(send_buf) < 0:
            self._logger.error('Error sending message')
    # end _send

# end class SandeshWriter


class SandeshSession(TcpSession):
    _KEEPALIVE_IDLE_TIME = 15  # in secs
    _KEEPALIVE_INTERVAL = 3  # in secs
    _KEEPALIVE_PROBES = 5
    _TCP_USER_TIMEOUT_OPT = 18
    _TCP_USER_TIMEOUT_VAL = 30000 # ms

    def __init__(self, sandesh_instance, server, event_handler,
                 sandesh_msg_handler):
        self._sandesh_instance = sandesh_instance
        self._logger = sandesh_instance._logger
        self._event_handler = event_handler
        self._reader = SandeshReader(self, sandesh_msg_handler)
        self._writer = SandeshWriter(self)
        self._send_queue = WorkQueue(self._send_sandesh,
                                     self._is_ready_to_send_sandesh)
        TcpSession.__init__(self, server)
    # end __init__

    # Public functions

    def sandesh_instance(self):
        return self._sandesh_instance
    # end sandesh_instance

    def is_send_queue_empty(self):
        return self._send_queue.is_queue_empty()
    # end is_send_queue_empty

    def is_connected(self):
        return self._connected
    # end is_connected

    def enqueue_sandesh(self, sandesh):
        self._send_queue.enqueue(sandesh)
    # end enqueue_sandesh

    def send_queue(self):
        return self._send_queue
    # end send_queue

    # Overloaded functions from TcpSession

    def connect(self):
        TcpSession.connect(self, timeout=5)
    # end connect

    def _on_read(self, buf):
        if self._reader.read_msg(buf) < 0:
            self._logger.error('SandeshReader Error. Close Collector session')
            self.close()
    # end _on_read

    def _handle_event(self, event):
        self._event_handler(self, event)
    # end _handle_event

    def _set_socket_options(self):
        self._socket.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
        if hasattr(socket, 'TCP_KEEPIDLE'):
            self._socket.setsockopt(
                socket.IPPROTO_TCP, socket.TCP_KEEPIDLE,
                self._KEEPALIVE_IDLE_TIME)
        if hasattr(socket, 'TCP_KEEPALIVE'):
            self._socket.setsockopt(
                socket.IPPROTO_TCP, socket.TCP_KEEPALIVE,
                self._KEEPALIVE_IDLE_TIME)
        if hasattr(socket, 'TCP_KEEPINTVL'):
            self._socket.setsockopt(
                socket.IPPROTO_TCP, socket.TCP_KEEPINTVL,
                self._KEEPALIVE_INTERVAL)
        if hasattr(socket, 'TCP_KEEPCNT'):
            self._socket.setsockopt(
                socket.IPPROTO_TCP, socket.TCP_KEEPCNT, self._KEEPALIVE_PROBES)
        try:
            self._socket.setsockopt(socket.IPPROTO_TCP,
                self._TCP_USER_TIMEOUT_OPT, self._TCP_USER_TIMEOUT_VAL)
        except:
            self._logger.error('setsockopt failed: option %d, value %d' %
                (self._TCP_USER_TIMEOUT_OPT, self._TCP_USER_TIMEOUT_VAL))
    # end _set_socket_options

    # Private functions

    def _send_sandesh(self, sandesh):
        if self._send_queue.is_queue_empty():
            more = False
        else:
            more = True
        if not self._connected:
            if self._sandesh_instance.is_logging_dropped_allowed(sandesh):
                self._logger.error(
                    "SANDESH: %s: %s" % ("Not connected", sandesh.log()))
            self._sandesh_instance.msg_stats().update_tx_drop_stats(
                sandesh.__class__.__name__, 0)
            return
        if sandesh.is_logging_allowed(self._sandesh_instance):
            self._logger.log(
                SandeshLogger.get_py_logger_level(sandesh.level()),
                sandesh.log())
        self._writer.send_msg(sandesh, more)
    # end _send_sandesh

    def _is_ready_to_send_sandesh(self):
        return self._sandesh_instance.is_send_queue_enabled()
    # end _is_ready_to_send_sandesh

# end class SandeshSession
